doc/SONiC_VRRP_HLD.md

```diff
@@ -0,0 +1,582 @@
+14) "True"
+15) "track_interface"
+16) "Ethernet7|weight|10,PortChannel001|weight|10,"
+
+### 4.2.2 APP_DB Changes
+
+**VRRP_TABLE**
+
+Producer:  VrrpSyncd 
+
+Consumer: VrrpOrch
+
+Description: This is a new table that contains VRRP state information. This entry will be added to APP_DB for each VIP for every Master VRRP instance. 
+
+Schema:
+
+```
+; New table
+; holds the VRRP state and VMAC information
+
+key = VRRP_TABLE:interface_name:vip:type
+    interface_name        ; interface name as a string. Vlan, Ethernet or PortChannel
+    vip                   ; virtual IP address in a.b.c.d/32 format
+    type                  ; IP address type string. only IPv4 supported currently
+    
+; field = value
+vmac = virtual_mac_address ; Virtual MAC address associated with VRRP instance
+```
+
+Example:- 
+
+**Key**: VRRP_TABLE:Vlan1000:[40.10.8.101/32:ipv4]
+
+**Value**: "vmac":"00:00:5e:00:01:08"
+
+## 4.3 Modules Design and Flows
+
+Below diagram depicts interactions of submodules for VRRP support.
+
+```mermaid
+sequenceDiagram
+    participant config_db
+    participant vrrpmgrd
+    participant keepalived
+    participant kernel
+    participant vrrpsyncd
+    participant app_db 
+    participant vrrporchd
+    participant asic_db
+
+    config_db->>vrrpmgrd: VRRP configuration changes
+    vrrpmgrd->>keepalived: Update configuration
+    Note over keepalived: Run VRRP state machine
+    keepalived->>kernel: Add/Del MACVLAN interface with virtual IP
+    kernel->>vrrpsyncd: Add/Del IP to MACVLAN interface
+    Note over vrrpsyncd: Add to APP_DB for Master instance
+    vrrpsyncd->>app_db: Add/Del interface, virtual IP
+    app_db->>vrrporchd: Listen to VRRP master/standby state
+    vrrporchd->>asic_db: Add/Del VMAC for VRRP interface 
+    
+
+
+```
+
+### vrrpmgrd
+
+- Listens to VRRP create, delete and parameter change in config_db. 
+- Upon change,
+  - Updates keepalived.conf file with new configuration
+  - Sends a signal to keepalived to read and apply new configuration from keepalived.conf file.
+
+### keepalived
+
+- Reads the configuration from keepalived.conf file.
+- Listens to kernel link and IP events.
+- Runs VRRP protocol state machine, 
+  - For VRRP instances state transition to Master programs a MACVLAN interface with VIP as interface IP in kernel
+  - For VRRP instances state transition to non-Master deletes MACVLAN interface, if programed, from kernel.
+
+### vrrpsyncd
+
+Listens to MACVLAN interface programming in kernel. Status of MACVLAN interface determines Master/Backup state of VRRP instances. VRRP_Table in APP_DB will be programmed with interface name and VIP for Master instances.
+
+- For every IP add to MACVLAN interface, adds interface name and Virtual IP in APP_DB in VRRP_Table
+- For every IP delete from MACVLAN interface, deletes interface name and Virtual IP in APP_DB from VRRP_Table
+
+### vrrporchd
+
+- Listens to VRRP_Table in APP_DB and adds virtual MAC entries in ASIC_DB for Master instances
```

Comment:
it is unclear. can you explain what does vrrporch do?

Reply:
vrrpsyncd listen to kernel and for each VRRP master instances, vrrpsyncd will add interface name, VIP & VMAC to APP_DB. 
vrrporch will inturn listen to APP_DB and for entry in VRRP_TABLE, program the VIP as my IP and the VMAC as my MAC(virtual RIF) in ASIC_DB.